### PR TITLE
KDECAN: rewrite driver to use CANSensor and add to AP_Periph

### PR DESCRIPTION
**Pull request by magicrub**
rewrite driver to use CANSensor and add to AP_Periph

### CODE REVIEW

**Comment by magicrub** [comment]
there's a few minor tweaks I need to apply to restrict it from compiling on non-vehicles and such

File: libraries/AP_KDECAN/AP_KDECAN.h [hunk 1]
```diff
@@ -20,23 +20,32 @@
  
 #pragma once
 
-#include <AP_CANManager/AP_CANDriver.h>
+#include <AP_HAL/AP_HAL.h>
 
-#if HAL_MAX_CAN_PROTOCOL_DRIVERS
+#ifndef HAL_KDECAN_ENABLED
```
Comment (peterbarker) [inline libraries/AP_KDECAN/AP_KDECAN.h:25]:
```suggestion
#ifndef AP_KDECAN_ENABLED
```

File: libraries/AP_KDECAN/AP_KDECAN.h [hunk 2]
```diff
@@ -20,23 +20,32 @@
...
-#include <AP_HAL/Semaphores.h>
+#if HAL_KDECAN_ENABLED
 
+#include <AP_CANManager/AP_CANSensor.h>
 #include <AP_Param/AP_Param.h>
 #include <AP_ESC_Telem/AP_ESC_Telem_Backend.h>
+#include <AP_HAL/utility/RingBuffer.h>
 
-#include <atomic>
+#define KDECAN_USE_EVENTS (CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS && CH_CFG_USE_EVENTS == TRUE)
```
Comment (peterbarker) [inline libraries/AP_KDECAN/AP_KDECAN.h:36]:
```suggestion
#define AP_KDECAN_USE_EVENTS (CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS && CH_CFG_USE_EVENTS == TRUE)
```

**Review by peterbarker — CHANGES_REQUESTED** [review]
As you say, there are per-vehicle problems here.  Last time I tried to make an `AP_KDECAN_ENABLED` define I discovered that there was a dependence on a motors library which not all vehicles supplied....

Sorry, didn't look at this in depth at all.  You've correctly tagged Sid :-)

**Comment by rmackay9** [comment]
We should be sure that this is tested on real hardware as well before merging.

**Comment by magicrub** [comment]
> We should be sure that this is tested on real hardware as well before merging.

I have two that I've tested with on the same network. What I haven't tested thoroughly is the enumeration stuff but I've tested it with how I think it's supposed to work. Would be nice to get someone else who uses KDECAN enumeration to test that

File: libraries/AP_KDECAN/AP_KDECAN.h (hunk 2 above)
Comment (magicrub) [reply libraries/AP_KDECAN/AP_KDECAN.h:36]: sure, that's a reasonable thing to do.

File: libraries/AP_KDECAN/AP_KDECAN.h (hunk 1 above)
Comment (magicrub) [reply libraries/AP_KDECAN/AP_KDECAN.h:25]: sure, that's a reasonable thing to do.

Comment (magicrub) [reply libraries/AP_KDECAN/AP_KDECAN.h]: fixed

File: libraries/AP_KDECAN/AP_KDECAN.h (hunk 2 above)
Comment (magicrub) [reply libraries/AP_KDECAN/AP_KDECAN.h]: fixed

**Comment by magicrub** [comment]
> Last time I tried to make an AP_KDECAN_ENABLED define I discovered that there was a dependence on a motors library which not all vehicles supplied....

I just added a self-defining enable in AP_Motors_Class.cpp so this sort of thing can get abstracted and detected at runtime. It causes a potential compile-order issue but it's at least working right now. Better solution would be for waf to detect if AP_Motors is included in the build and then set the define.. but this already has a bit of feature-creep as it is

Now my KDECAN has this in the arming check:
#if defined(AP_MOTORS_ENABLE) && AP_MOTORS_ENABLE == 1

**Comment by magicrub** [comment]
I've pulled out some AP_Perioph infrastructure changes into it's own PR https://github.com/ArduPilot/ardupilot/pull/19672 so we can see more of what adding KDECAN is will less feature-creep infrastructure changes

**Comment by magicrub** [comment]
closing in favor of https://github.com/ArduPilot/ardupilot/pull/23238

**Pull request closed without merging**